output/cloud/expv2/output.go

```diff
@@ -0,0 +1,265 @@
+// Package expv2 contains a Cloud output using a Protobuf
+// binary format for encoding payloads.
+package expv2
+
+import (
+	"context"
+	"time"
+
+	"github.com/mstoykov/atlas"
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/metrics"
+	"go.k6.io/k6/output"
+	"go.k6.io/k6/output/cloud/expv2/pbcloud"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// TestName is the default Cloud test name
+const (
+	TestName = "k6 test"
+)
+
+// Output sends result data to the k6 Cloud service.
+type Output struct {
+	output.SampleBuffer
+
+	config      cloudapi.Config
+	referenceID string
+
+	logger          logrus.FieldLogger
+	client          *MetricsClient
+	periodicFlusher *output.PeriodicFlusher
+
+	// TODO: if the metric refactor (#2905) will introduce
+	// a sequential ID for metrics
+	// then we could reuse the same strategy here
+	activeSeries map[*metrics.Metric]aggregatedSamples
+}
+
+// New creates a new cloud output.
+func New(logger logrus.FieldLogger, conf cloudapi.Config) (*Output, error) {
+	mc, err := NewMetricsClient(logger, conf.Host.String, conf.Token.String)
+	if err != nil {
+		return nil, err
+	}
+	return &Output{
+		config:       conf,
+		client:       mc,
+		logger:       logger.WithFields(logrus.Fields{"output": "cloudv2"}),
+		activeSeries: make(map[*metrics.Metric]aggregatedSamples),
+	}, nil
+}
+
+// Start starts the output.
+func (o *Output) Start() error {
+	o.logger.Debug("Starting...")
+
+	// TODO: merge here the part executed by v1 when we will drop it
+	pf, err := output.NewPeriodicFlusher(o.config.MetricPushInterval.TimeDuration(), o.flushMetrics)
+	if err != nil {
+		return err
+	}
+	o.logger.Debug("Started!")
+	o.periodicFlusher = pf
+	return nil
+}
+
+// StopWithTestError stops the output.
+func (o *Output) StopWithTestError(testErr error) error {
+	o.logger.Debug("Stopping...")
+	defer o.logger.Debug("Stopped!")
+	o.periodicFlusher.Stop()
+	return nil
+}
+
+// SetReferenceID sets the Cloud's test ID.
+func (o *Output) SetReferenceID(refID string) {
```

Comment:
Because the reference ID is generated when we _Start_ the output and not when we instantiate the output. The reference id is generated from the cloud backend.